### PR TITLE
Create MinLeverage control

### PR DESCRIPTION
**Pull request by srklonaris**
Similar to `MaxLeverage` create a `MinLeverage` control.  

`MinLeverage` takes a float as well as a deadline date.  `MinLeverage` validations will only happen after the algorithm has reached the deadline date.  This allows algorithms enough time to reach their min leverage because it may take time for them to have a non zero leverage.

### CODE REVIEW

File: zipline/finance/controls.py [hunk 1]
```diff
@@ -425,3 +425,48 @@ def validate(self,
...
+
+class MinLeverage(AccountControl):
+    """
+    AccountControl representing a limit on the minimum leverage allowed
+    by the algorithm after a threshold period of time.
+    """
+
+    def __init__(self, min_leverage, deadline):
+        """
+        max_leverage is the gross leverage in decimal form. For example,
```
Comment (ssanderson) [inline zipline/finance/controls.py]: Docstring here is still copied from `MaxLeverage`.  We've also standardized on [numpy-style docstrings](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt) in zipline since `MaxLeverage` was written.  I don't think we need to go back and change existing code, but for new docstrings we should still use the style we use elsewhere in Zipline.

File: zipline/finance/controls.py [hunk 2]
```diff
@@ -425,3 +425,48 @@ def validate(self,
...
+        2, limits an algorithm to trading at most double the account value.
+        """
+        super(MinLeverage, self).__init__(min_leverage=min_leverage,
+                                          deadline=deadline)
+        self.min_leverage = min_leverage
+        self.deadline = deadline
+
+        if min_leverage is None:
```
Comment (ssanderson) [inline zipline/finance/controls.py]: You might want to look at using the decorators in `zipline.utils.input_validation` to implement these checks. I think `expect_types(min_leverage=(int, float), deadline=datetime)` and `expect_strictly_bounded(min_leverage=(0, None))` would capture the behavior we want here.  If you do end up using those functions, you probably want to pass `__funcname='MinLeverage'` so that the generated error messages reference `MinLeverage` instead of `__init__`.

File: zipline/test_algorithms.py [hunk 3]
```diff
@@ -485,6 +485,11 @@ def initialize(self, max_leverage=None):
         self.set_max_leverage(max_leverage=max_leverage)
 
 
+class SetMinLeverageAlgorithm(TradingAlgorithm):
+    def initialize(self, min_leverage=None, deadline=None):
```
Comment (ssanderson) [inline zipline/test_algorithms.py]: Should we just make these required if it's invalid to pass `None` for these values to `set_min_leverage`?

File: zipline/algorithm.py [hunk 4]
```diff
@@ -2234,6 +2235,22 @@ def set_max_leverage(self, max_leverage):
...
+    def set_min_leverage(self, min_leverage, deadline):
+        """
+        Set a limit on the minimum leverage of the algorithm.
+
+        Parameters
+        ----------
+        min_leverage : float
+            The minimum leverage for the algorithm. If not provided there will
+            be no minimum.
+        deadline : datetime
```
Comment (ssanderson) [inline zipline/algorithm.py]: This might be more nicely expressed as a timedelta from the start of the algorithm instead of a datetime.

File: zipline/algorithm.py [hunk 5]
```diff
@@ -2234,6 +2235,22 @@ def set_max_leverage(self, max_leverage):
...
 
+    @api_method
+    def set_min_leverage(self, min_leverage, deadline):
+        """
+        Set a limit on the minimum leverage of the algorithm.
+
+        Parameters
+        ----------
+        min_leverage : float
+            The minimum leverage for the algorithm. If not provided there will
```
Comment (ssanderson) [inline zipline/algorithm.py]: Is the second sentence here actually true?  If not, should we just remove it?

File: zipline/finance/controls.py [hunk 6]
```diff
@@ -425,3 +425,48 @@ def validate(self,
...
+                "min_leverage must be positive"
+            )
+
+        if deadline is None:
+            raise ValueError(
+                "Must supply deadline"
+            )
+
+    def validate(self,
+                 _portfolio,
```
Comment (ssanderson) [inline zipline/finance/controls.py]: The underscores here are a weird circa-2014 scott-ism: the contention elsewhere in this file is that the paramters start with an underscore if they're unused in the function, which was a convention I'd used in a previous codebase.  We don't do that anywhere else in zipline though.  I'd either follow the existing convention in this file (in which case `algo_datetime` and `account` shouldn't have underscores), or just make them all normal-looking parameters.

File: tests/test_algorithm.py [hunk 7]
```diff
@@ -3531,6 +3532,42 @@ def handle_data(algo, data):
                                        env=self.env)
         self.check_algo_succeeds(algo, handle_data)
 
+    def test_set_min_leverage(self):
+        # Algo date is not past deadline so algorithm succeeds
+        deadline = pd.Timestamp('2010-01-05', tz='UTC')
+
+        # Set min leverage to 0 so buying one share fails.
+        def handle_data(algo, data):
```
Comment (ssanderson) [inline tests/test_algorithm.py]: I think this `handle_data` definition is sufficient for all the tests in this function.  Seems like it'd be simpler and clearer to just always use this instead of re-defining it for each test?

File: tests/test_algorithm.py [hunk 8]
```diff
@@ -3531,6 +3532,42 @@ def handle_data(algo, data):
                                        env=self.env)
         self.check_algo_succeeds(algo, handle_data)
 
+    def test_set_min_leverage(self):
+        # Algo date is not past deadline so algorithm succeeds
+        deadline = pd.Timestamp('2010-01-05', tz='UTC')
+
+        # Set min leverage to 0 so buying one share fails.
```
Comment (ssanderson) [inline tests/test_algorithm.py]: This comment doesn't match the test case.

File: tests/test_algorithm.py [hunk 9]
```diff
@@ -3531,6 +3532,42 @@ def handle_data(algo, data):
                                        env=self.env)
         self.check_algo_succeeds(algo, handle_data)
 
+    def test_set_min_leverage(self):
+        # Algo date is not past deadline so algorithm succeeds
```
Comment (ssanderson) [inline tests/test_algorithm.py]: What is "algo date" referring to in this comment?

File: zipline/test_algorithms.py [hunk 10]
```diff
@@ -485,6 +485,11 @@ def initialize(self, max_leverage=None):
         self.set_max_leverage(max_leverage=max_leverage)
 
 
+class SetMinLeverageAlgorithm(TradingAlgorithm):
+    def initialize(self, min_leverage=None, timedelta=None):
```
Comment (ssanderson) [inline zipline/test_algorithms.py]: Should these just be required if passing them as None will isn't actually supported?

File: zipline/algorithm.py [hunk 11]
```diff
@@ -2234,6 +2235,23 @@ def set_max_leverage(self, max_leverage):
         control = MaxLeverage(max_leverage)
         self.register_account_control(control)
 
+    @api_method
+    def set_min_leverage(self, min_leverage, timedelta):
```
Comment (ssanderson) [inline zipline/algorithm.py]: I would still call this `deadline`.  `timedelta` is shadowing the name of a built-in type.

File: tests/test_algorithm.py [hunk 12]
```diff
@@ -3531,6 +3532,42 @@ def handle_data(algo, data):
...
+        offset = pd.Timedelta('10 days')
+
+        # Set min leverage to 0 so buying one share fails.
```
Comment (ssanderson) [inline tests/test_algorithm.py]: This comment doesn't match the test case.

File: tests/test_algorithm.py [hunk 13]
```diff
@@ -3531,6 +3532,42 @@ def handle_data(algo, data):
...
+        def handle_data(algo, data):
+            algo.order(algo.sid(self.sidint), 1)
+            algo.record(latest_time=algo.get_datetime())
+
+        algo = SetMinLeverageAlgorithm(1, offset, sim_params=self.sim_params,
+                                       env=self.env)
+        self.check_algo_fails(algo, handle_data)
+        self.assertEqual(
+            algo.recorded_vars['latest_time'],
+            pd.Timestamp('2006-01-04 21:00:00', tz='UTC'),
```
Comment (ssanderson) [inline tests/test_algorithm.py]: Can we also test that this fails at 01-05 if we increase the offset delta to 2 days?  (I might just break out those two cases into separate test methods instead of having all the cases at once).

File: tests/test_algorithm.py [hunk 14]
```diff
@@ -3531,6 +3532,42 @@ def handle_data(algo, data):
                                        env=self.env)
         self.check_algo_succeeds(algo, handle_data)
 
+    def test_set_min_leverage(self):
+        # Algo date is not past deadline so algorithm succeeds
+        offset = pd.Timedelta('10 days')
+
+        # Set min leverage to 0 so buying one share fails.
+        def handle_data(algo, data):
```
Comment (ssanderson) [inline tests/test_algorithm.py]: All the `handle_data` functions we define in this method are functionally the same.  Is there a reason we're defining a separate one each time?

File: tests/test_algorithm.py [hunk 15]
```diff
@@ -3531,6 +3532,42 @@ def handle_data(algo, data):
...
+            pd.Timestamp('2006-01-04 21:00:00', tz='UTC'),
+        )
+
+        # Set min leverage to 0 so buying one share passes
+        def handle_data(algo, data):
+            algo.order(algo.sid(self.sidint), 1)
+
+        algo = SetMinLeverageAlgorithm(0, offset, sim_params=self.sim_params,
+                                       env=self.env)
+        self.check_algo_succeeds(algo, handle_data)
```
Comment (ssanderson) [inline tests/test_algorithm.py]: I think we need a test case where we set a minimum leverage amount and the algorithm has a leverage above that limit after the deadline and doesn't crash.  Bonus points if the algorithm then goes below the leverage limit and crashes.

Comment (ssanderson) [reply tests/test_algorithm.py]: This might be easier to implement using `order_target_percent` to hit fixed leverage amounts instead of ordering fixed numbers of shares.

File: zipline/algorithm.py (hunk 11 above)
Comment (srklonaris) [reply zipline/algorithm.py]: What about `offset`, since `deadline` is what is computed and passed into `MinLeverage`

Comment (ssanderson) [reply zipline/algorithm.py]: As a consumer of this function, I don't think `offset` would be a very clear name for this parameter.  My immediate question would be "offset from what"?  Maybe `grace_period`?

File: tests/test_algorithm.py [hunk 16]
```diff
@@ -3531,6 +3532,46 @@ def handle_data(algo, data):
                                        env=self.env)
         self.check_algo_succeeds(algo, handle_data)
 
+    def test_set_min_leverage(self):
+        def handle_data(algo, data):
+            algo.order_target_percent(algo.sid(self.sidint), .5)
+            algo.record(latest_time=algo.get_datetime())
+
+        # Set min leverage to 1.
```
Comment (ssanderson) [inline tests/test_algorithm.py]: I don't think this comment matches the test case?

File: tests/test_algorithm.py [hunk 17]
```diff
@@ -3531,6 +3532,46 @@ def handle_data(algo, data):
...
+        # The algorithm will succeed because it doesn't run for more
+        # than 10 days.
+        offset = pd.Timedelta('10 days')
+        algo = SetMinLeverageAlgorithm(0, offset, sim_params=self.sim_params,
+                                       env=self.env)
+        self.check_algo_succeeds(algo, handle_data)
+
+        # The algorithm will fail because it doesn't reach a min leverage of 1
```
Comment (ssanderson) [inline tests/test_algorithm.py]: :+1:

File: tests/test_algorithm.py [hunk 18]
```diff
@@ -3531,6 +3532,46 @@ def handle_data(algo, data):
...
+    def test_set_min_leverage(self):
+        def handle_data(algo, data):
+            algo.order_target_percent(algo.sid(self.sidint), .5)
+            algo.record(latest_time=algo.get_datetime())
+
+        # Set min leverage to 1.
+        # The algorithm will succeed because it doesn't run for more
+        # than 10 days.
+        offset = pd.Timedelta('10 days')
+        algo = SetMinLeverageAlgorithm(0, offset, sim_params=self.sim_params,
```
Comment (ssanderson) [inline tests/test_algorithm.py]: I think this should be a 1?

File: tests/test_algorithm.py [hunk 19]
```diff
@@ -3531,6 +3532,46 @@ def handle_data(algo, data):
...
+        offset = pd.Timedelta('1 days')
+        algo = SetMinLeverageAlgorithm(1, offset, sim_params=self.sim_params,
+                                       env=self.env)
+        self.check_algo_fails(algo, handle_data)
+        self.assertEqual(
+            algo.recorded_vars['latest_time'],
+            pd.Timestamp('2006-01-04 21:00:00', tz='UTC'),
+        )
+
+        # Increase the offset to 2 days, and the algorithm fails a day later
```
Comment (ssanderson) [inline tests/test_algorithm.py]: This is nicely clear to follow as a reader :+1: .

File: zipline/algorithm.py [hunk 20]
```diff
@@ -2234,6 +2235,22 @@ def set_max_leverage(self, max_leverage):
...
+    def set_min_leverage(self, min_leverage, grace_period):
+        """
+        Set a limit on the minimum leverage of the algorithm.
+
+        Parameters
+        ----------
+        min_leverage : float
+            The minimum leverage for the algorithm.
+        grace_period : pd.Timedelta
+            The offset from the start date used to enforce a minimum leverage
```
Comment (ssanderson) [inline zipline/algorithm.py]: Missing period here.

File: zipline/finance/controls.py [hunk 21]
```diff
@@ -425,3 +430,40 @@ def validate(self,
...
+class MinLeverage(AccountControl):
+    """
+    AccountControl representing a limit on the minimum leverage allowed
+    by the algorithm after a threshold period of time.
+    """
+
+    @expect_types(
+        __funcname='MinLeverage',
+        min_leverage=(int, float),
+        deadline=pd.datetime
```
Comment (ssanderson) [inline zipline/finance/controls.py]: `pd.datetime` is just an alias of the standard library `datetime.datetime` (presumably b/c something in the top level pandas module does `from datetime import datetime`).  We should use the standard library location here.

File: zipline/finance/controls.py [hunk 22]
```diff
@@ -425,3 +430,40 @@ def validate(self,
...
+    )
+    @expect_bounded(__funcname='MinLeverage', min_leverage=(0, None))
+    def __init__(self, min_leverage, deadline):
+        """
```
Comment (ssanderson) [inline zipline/finance/controls.py]: We generally don't put docstrings on `__init__` functions; we use the class docstring documenting the parameters to a class and describing what it's for.

File: zipline/finance/controls.py [hunk 23]
```diff
@@ -425,3 +430,40 @@ def validate(self,
         """
         if _account.leverage > self.max_leverage:
             self.fail()
+
+
+class MinLeverage(AccountControl):
+    """
```
Comment (ssanderson) [inline zipline/finance/controls.py]: This should probably also be a numpy-style docstring.

**Comment by ssanderson** [comment]
LGTM
